pkg/embed/append.go

```diff
@@ -0,0 +1,175 @@
+	if _, err = fp.Read(buf); err != nil {
+		return 0, 0, fmt.Errorf("unable to read end of file: %w", err)
+	}
+	if !strings.HasPrefix(string(buf), EndMark) {
+		return 0, 0, nil
+	}
+	strbuf := strings.TrimPrefix(string(buf), EndMark)
+	size, err := strconv.ParseInt(strbuf, 10, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("unable to parse end of file: %w", err)
+	}
+	rewind := size + int64(ruler.Len())
+	return rewind, size, nil
+}
+
+// ReadEmbeddedKotsRelease reads the embedded release from the given file path.
+// Returns the embedded data as it was embedded. If no embedded data is found,
+// returns nil and no error.
+func ReadEmbeddedKotsRelease(bin string) (io.Reader, error) {
+	rewind, size, err := CalculateRewind(bin)
+	if err != nil {
+		return nil, fmt.Errorf("unable to calculate rewind: %w", err)
+	} else if rewind == 0 {
+		return nil, nil
+	}
+	fp, err := os.Open(bin)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open file %s: %w", bin, err)
+	}
+	defer fp.Close()
+	if _, err := fp.Seek(-int64(rewind), io.SeekEnd); err != nil {
+		return nil, fmt.Errorf("unable to seek to end of file: %w", err)
+	}
+	data := make([]byte, size)
+	if _, err := fp.Read(data); err != nil {
+		return nil, fmt.Errorf("unable to read end of file: %w", err)
+	}
+	return bytes.NewReader(data), nil
+}
+
+// ReadAndParseEmbeddedKotsRelease reads the embedded release from the given file path.
+// Parses the release into an afero.FS and returns it. If no embedded data is found,
+// returns nil as afero.FS and no error.
+func ReadAndParseEmbeddedKotsRelease(bin string) (afero.Fs, error) {
+	fp, err := ReadEmbeddedKotsRelease(bin)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read embedded data: %w", err)
+	} else if fp == nil {
+		return nil, nil
+	}
+	gzreader, err := gzip.NewReader(fp)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create gzip reader: %w", err)
+	}
+	defer gzreader.Close()
+	tarreader := tar.NewReader(gzreader)
+	fs := afero.NewMemMapFs()
+	for {
+		header, err := tarreader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to read tar header: %w", err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			err = fs.MkdirAll(header.Name, os.FileMode(header.Mode))
+			if err != nil {
+				return nil, fmt.Errorf("unable to create dir: %w", err)
+			}
+		case tar.TypeReg:
+			outfile, err := fs.Create(header.Name)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create file: %w", err)
+			}
+			defer outfile.Close()
+			if _, err = io.Copy(outfile, tarreader); err != nil {
+				return nil, fmt.Errorf("unable to copy file: %w", err)
+			}
+		}
+	}
+	return fs, nil
+}
+
+// Build builds the binary with the embedded release. Returns an io.ReadCloser
+// from where the binary can be read. Returns the ReadCloser, the total size of
+// the binary and an error if any.
+func Build(bin string, release string) (io.ReadCloser, int64, error) {
```

Comment:
I would _prefer_ to have the inputs be readclosers here too, that way this can be entirely streamed and not require writing to disk (we would then be able to skip the airgap builder service)

Reply:
We would have to get the sizes with a wrapper
```
type wrapper struct {
    io.Reader
    n int
}

func (w *wrapper) Read(p []byte) (int, error) {
    n, err := w.Reader.Read(p)
    w.n += n
    return n, err
}
```